### ISSUE TITLE
Yasson requires its own/internal JSON-P implementation to work ... ???

### ISSUE THREAD

**Issue by amoscatelli** [eclipse-ee4j/yasson#102]
I am implementing polymorphism mechanism with JSON-B, similar to Jackson one.
So I need to scan for a "@type" attribute and then deserialize again with the correct class.

`public class IdentifiableSerializer implements JsonbDeserializer<FindCriteriaDTO> {

    private final Reflections reflections;

    public IdentifiableSerializer(Reflections reflections) {
        this.reflections = reflections;
    }
    
    @Override
    public FindCriteriaDTO deserialize(JsonParser parser, DeserializationContext ctx, Type rtType) {
        
        JsonObject preview = ctx.deserialize(JsonObject.class, parser);
        
        if (preview.keySet().contains("@type")){
            Optional<Class<? extends FindCriteriaDTO>> type = reflections.getSubTypesOf(FindCriteriaDTO.class).stream().filter(e -> StringUtils.equals(preview.getString("@type"), e.getSimpleName())).findFirst();
            
            if (type.isPresent()){
                
                StringWriter stringWriter = new StringWriter();
                Json.createWriter(stringWriter).writeObject(preview);
                JsonParser newParser = Json.createParser(new StringReader(stringWriter.toString()));
                
                FindCriteriaDTO deserialize = ctx.deserialize(type.get(), newParser);
                
                return deserialize;
            }
            
        }
        
        return ctx.deserialize(FindCriteriaDTO.class, parser);
        
    }`

If I create a new JsonParser calling the JSON-P method a different implementation is returned (and this should not be an issue! json-b and json-p are different specifications!). But then I see this exception :

org.glassfish.json.JsonParserImpl cannot be cast to org.eclipse.yasson.internal.JsonbParser

This casting is the reason (org.eclipse.yasson.internal.serializer.AbstractContainerDeserializer:58) :

` @Override
    public final T deserialize(JsonParser parser, DeserializationContext context, Type rtType) {
        Unmarshaller ctx = (Unmarshaller) context;
        ctx.setCurrent(this);
        deserializeInternal((JsonbParser) parser, ctx);
        ctx.setCurrent(getWrapper());
        return getInstance((Unmarshaller) context);
    }`

I don't think this is the proper behavior.
Am I wrong ?

**Comment by m0mus** [comment]
@bravehorsie please help

**Comment by bravehorsie** [comment]
You should not create it. It is passed in the deserialize method parameter. JsonbParser decorates to Jsonp, doing its business around it.

**Comment by amoscatelli** [comment]
I NEED to create it. And the spec says the deserializer context should accept a standard JsonParser, not a org.eclipse.yasson.internal.JsonbParser. This is what interfaces stand for.

This is clearly out of specification.

**Comment by bravehorsie** [comment]
Specification refers to jsonp api, not the RI. JsonParser is what you get in (de)serializer to drive it yourself. To deserialize java object from json you need to create Jsonb runtime instance, and call `jsonb.deserialize(json, FindCriteriaDTO.class)`. You have both in your deserializer implementation code.

**Comment by amoscatelli** [comment]
Yea and you are not using the JSONP api, but your specific implementation.

JSONP API:
javax.json.stream.JsonParser
Provides forward, read-only access to JSON data in a streaming way.

JSONB API :
javax.json.bind.serializer.DeserializationContext<T>
Provides JSONB Mapper functionality on top of JSONP parser
<T> T deserialize(Class<T> clazz, JsonParser parser);

DeserializationContext deserialize method should work with ANY JsonParser jsonp implementation. Not only RI (org.glassfish.json.JsonParserImpl), not only others (org.eclipse.yasson.internal.JsonbParser). It should work independently of what implementation the API Json.createParser() could return.

Regarding your solution/workaround you proposed I do NOT have the reference the jsonb inside a JsonbDeserializer. I could inject it in some way, but I do not think that's the clean way.

**Comment by bravehorsie** [comment]
> DeserializationContext deserialize method should work with ANY JsonParser jsonp implementation. Not only RI (org.glassfish.json.JsonParserImpl), not only others (org.eclipse.yasson.internal.JsonbParser). It should work independently of what implementation the API Json.createParser() could return.

So it does. You can provide any JSONP implementation to Yasson to work with it. What you do is you substitute running JSONP parser which has a cursor at some point of the document with new parser instance reading new document but you call current runtime on it, that is wrong. If you want to read a new document you should create new runtime with your parser instance.

**Comment by dylanBehetre** [comment]
> So it does.
No, it isn't, if we put an other JsonParser (which is legit and created in standard way) in parameter of the DeserializationContext.deserialize(Class<T> clazz, JsonParser parser) method, it throw an Exception `Internal error: org.glassfish.json.JsonParserImpl cannot be cast to org.eclipse.yasson.internal.JsonbParser` because your implementation of the deserialize method cast the parser parameter as `JsonbParser`. But the contract in the interface say `JsonParser ` and not `JsonbParser`. So you are wrong. It's all.

> What you do is you substitute running JSONP parser which has a cursor at some point of the document with new parser instance reading new document but you call current runtime on it, that is wrong. If you want to read a new document you should create new runtime with your parser instance.

Who said that is wrong ?
Firstly, he not substitute the running JSONP parser, he only used it. 
Secondly, the purpose of the `public final T deserialize(JsonParser parser, DeserializationContext context, Type rtType)` method is `Drives JSONP {@link JsonParser} to deserialize json document.`, so, if we need to duplicate the parser to process the structure two time, why not ? I don't understand.

**Comment by dylanBehetre** [comment]
We can wrapped the JsonParser in a JsonbParser by using `JsonbRiParser `or `UserDeserializerParser`. But which is more adapted ?

**Comment by alessandro-tucci-visiontech** [comment]
> So it does. You can provide any JSONP implementation to Yasson to work with it.

I am not getting how a custom JSONP implementation can be actually provided to Yasson. From what I can see from here: https://github.com/eclipse-ee4j/yasson/blob/f8263f765368af3609e55da568f596a95483cbea/src/main/java/org/eclipse/yasson/internal/serializer/AbstractContainerDeserializer.java#L64
there is an explicit cast to **org.eclipse.yasson.internal.JsonbParser** of the provided JsonParser parameter. So, from what I see, I would say that there is no way to avoid using the Yasson parser, currently.
Am I wrong?
